File: Base/Queue/Javascript/CircleQueue.js
class CircleQueue {
  constructor() {
    this.SIZE = 100;
    this.head = 0;
    this.tail = 0;
    this.queue = [];
  }

  enqueue(item) {
    // 队列已满
    if ((this.tail + 1) % this.SIZE === this.head) {
      return false;
    }

    this.queue[this.tail++] = item;
    this.tail = this.tail % this.SIZE;
    return true;
  }

  dequeue() {
    if (this.head === this.tail) {
      return null;
    }

    let item = this.queue[this.head++];
    this.head = this.head % this.SIZE;
    return item;
  }

  print() {
    let temp = [];
    for (let i = this.head; i < this.tail; i++) {
      temp.push(this.queue[i]);
    }
    console.log(temp);
  }
}

let circle_queue = new CircleQueue();
circle_queue.enqueue(1);
circle_queue.enqueue(2);
circle_queue.enqueue(3);
circle_queue.print();
console.log('dequeue', circle_queue.dequeue());
circle_queue.print();

File: Queue/Rust/src/circle_queue.rs
use std::fmt::{Debug, Display};

#[derive(Debug)]
pub struct CircleQueue<T> {
    maxsize: usize,
    head: usize,
    tail: usize,
    queue: Vec<T>,
}

impl<T> CircleQueue<T> {
    pub fn new() -> CircleQueue<T> {
        CircleQueue {
            maxsize: 100,
            head: 0,
            tail: 0,
            queue: Vec::with_capacity(100),
        }
    }

    pub fn enqueue(&mut self, item: T) -> bool {
        if (self.tail + 1) % self.maxsize == self.head {
            return false;
        }
        self.queue.push(item);
        self.tail = (self.tail + 1) % self.maxsize;

        true
    }

    pub fn dequeue(&mut self) -> Option<&T> {
        if self.head == self.tail {
            return None;
        }
        let data = &self.queue[self.head];
        self.head = (self.head + 1) % self.maxsize;
        Option::Some(data)
    }
}

impl<T: Display + Debug> CircleQueue<T> {
    pub fn print(&self) {
        let mut data: Vec<&T> = Vec::new();
        for i in self.head..self.tail {
            let value = &self.queue[i];
            data.push(value);
        }
        println!("circle_queue = {:?}", data);
    }
}